{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Probe Designer \n",
    "## for RNA of specific genes"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# basci env\n",
    "import os\n",
    "import pandas as pd\n",
    "import time\n",
    "import json\n",
    "\n",
    "# data process of file from ncbi\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "# get gene data from ncbi\n",
    "from Bio import Entrez\n",
    "\n",
    "# blast and xml file process\n",
    "from Bio.Blast import NCBIWWW\n",
    "from Bio.Blast import NCBIXML\n",
    "\n",
    "# dir\n",
    "os.chdir('C:/Users/14187/Documents/Github/Probe_designer/example_dataset_immune')\n",
    "current_time = time.localtime()\n",
    "formatted_time = time.strftime(\"%Y%m%d_%H%M%S\", current_time)\n",
    "tmp = './results/' + formatted_time + '/tmp/'\n",
    "output = './results/' + formatted_time + '/'\n",
    "# tmp = './results/' + '20230425_103142' + '/tmp/'\n",
    "# output = './results/' + '20230425_103142' + '/'\n",
    "pre_binding_dir = tmp + 'pre_binding/'\n",
    "try:os.makedirs(tmp)\n",
    "except:pass\n",
    "\n",
    "# basic vatiables\n",
    "gene_name_list_file = 'gene_name_list.txt'\n",
    "gene_name_list_tosearch = 'gene_name_list_tosearch.txt'\n",
    "pre_binding_file_suffix = '_pre_binding.fasta'\n",
    "total_pre_binding_file_name = '_total_pre_binding.fasta'\n",
    "gene_id_name_file = '1_id_list.txt'\n",
    "gene_seq_in_file = '2_gene_seq_in_file.gb'\n",
    "pre_binding_num_file = '3_pre_binding_num.json'\n",
    "blast_results_file = '4_blast_results.xml'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "binding_site_FOIs = ['accession', 'gene_name', 'mol_type', 'organism', 'binding', 'wanted']\n",
    "align_FOIs = ['align_num', 'align_accession', 'align_descrip', 'plus/minus']\n",
    "# create the dataframe of Field of interest\n",
    "FOI = pd.DataFrame(columns=binding_site_FOIs+\n",
    "                   align_FOIs)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### get gene from ncbi dataset\n",
    "https://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EFetch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get gene id and other information from ncbi dataset(api)\n",
    "# generate gene_search_list from gene_name_list\n",
    "organism_of_interest = 'Mus musculus'\n",
    "n_type_of_interest = 'mRNA'\n",
    "with open(gene_name_list_file) as f:\n",
    "    gene_name_list = f.read().splitlines()\n",
    "gene_search_list = [', '.join([name, organism_of_interest, n_type_of_interest]) for name in gene_name_list]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "id_list = []\n",
    "for gene_search in gene_search_list:\n",
    "    Entrez.email = \"1418767067@qq.com\"\n",
    "    handle = Entrez.esearch(db=\"nuccore\", term=gene_search)\n",
    "    record = Entrez.read(handle)\n",
    "    handle.close()\n",
    "    id_list += record[\"IdList\"][:1]\n",
    "with open(tmp+gene_id_name_file, 'w') as f:\n",
    "    f.write('\\n'.join(id_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(tmp+gene_id_name_file, 'r') as f:\n",
    "    id_list = f.read().split('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get a genbank file to get other information about the genes\n",
    "fetch_per_round = 3\n",
    "round = len(id_list) // fetch_per_round\n",
    "for i in range(round):\n",
    "    id_list_per_round = id_list[i*fetch_per_round: (i+1)*fetch_per_round]\n",
    "    Entrez.email = \"1418767067@qq.com\"\n",
    "    handle = Entrez.efetch(db=\"nuccore\", strand=1,  # plus if strand=1\n",
    "                        id=id_list_per_round, rettype=\"gbwithparts\", retmode=\"text\")\n",
    "    seq_record = handle.read()\n",
    "    handle.close()\n",
    "    print(i+1, '{:.2f} %'.format((i+1)/round*100))\n",
    "    with open(tmp+gene_seq_in_file, 'a') as f:\n",
    "        f.write(seq_record)\n",
    "        \n",
    "id_list_per_round = id_list[(i+1)*fetch_per_round:]\n",
    "Entrez.email = \"1418767067@qq.com\"\n",
    "handle = Entrez.efetch(db=\"nuccore\", strand=1,  # plus if strand=1\n",
    "                        id=id_list_per_round, rettype=\"gbwithparts\", retmode=\"text\")\n",
    "seq_record = handle.read()\n",
    "handle.close()\n",
    "print(i+1, '{:.2f} %'.format((i+1)/round*100))\n",
    "\n",
    "with open(tmp+gene_seq_in_file, 'a') as f:\n",
    "    f.write(seq_record)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### binding site searcher"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# search on gene sequence to get binding site(more than need)\n",
    "file_in = tmp + gene_seq_in_file\n",
    "file_out_dir = pre_binding_dir\n",
    "try:\n",
    "    os.mkdir(file_out_dir)\n",
    "except:\n",
    "    pass\n",
    "\n",
    "BDS_len = 40\n",
    "max_num = 35\n",
    "\n",
    "pre_binding_num = {}\n",
    "pos, length = 0, 0\n",
    "for gene_seq_in in SeqIO.parse(tmp+gene_seq_in_file, 'genbank'):\n",
    "    # get information of gene\n",
    "    id = gene_seq_in.id # get seq id\n",
    "    # get gene_name\n",
    "    if gene_seq_in.features:\n",
    "        for feature in gene_seq_in.features:\n",
    "            if feature.type == \"CDS\":\n",
    "                gene_name = feature.qualifiers.get(\"gene\", ['NAN'])[0]\n",
    "    # get molecule_type\n",
    "    mol_type = gene_seq_in.annotations['molecule_type']\n",
    "    # get organism\n",
    "    organism = gene_seq_in.annotations['organism'] \n",
    "    # get minus seq\n",
    "    translib = {'A':'T','T':'A','C':'G','G':'C'}\n",
    "    try: \n",
    "        seq_minus = [translib[i] for i in str(gene_seq_in.seq)]\n",
    "        seq = ''.join(list(reversed(seq_minus)))\n",
    "    except: seq = str(gene_seq_in.seq)\n",
    "    \n",
    "\n",
    "    # set start point and pre_binding_num\n",
    "    length = len(seq)\n",
    "    pre_binding_num_tmp = min(length//BDS_len, max_num)\n",
    "    st = length // 2 - pre_binding_num_tmp * BDS_len // 2\n",
    "    # generate and write pre_binding for each gene in a fasta file\n",
    "    record_list = []\n",
    "    file_out = file_out_dir + id + pre_binding_file_suffix\n",
    "    for i in range(pre_binding_num_tmp):\n",
    "        pre_binding_tmp = seq[st+i*BDS_len:st+i*BDS_len+BDS_len]\n",
    "        record_list.append(SeqRecord(\n",
    "            Seq(pre_binding_tmp),\n",
    "            id='pre_binding'+str(i), \n",
    "            description='|'.join([id,gene_name,organism,mol_type])\n",
    "        ))\n",
    "        # add information about binding sites to FOI\n",
    "        FOI.loc[i+pos, binding_site_FOIs] \\\n",
    "             = [id, gene_name, mol_type, organism, pre_binding_tmp, True]\n",
    "        \n",
    "    pos += pre_binding_num_tmp\n",
    "    # write pre_binding to files\n",
    "    with open(file_out, \"w\") as f:\n",
    "        for new_record in record_list:\n",
    "            SeqIO.write(new_record, f, \"fasta\")\n",
    "    with open(file_out_dir+total_pre_binding_file_name, \"a\") as handle:\n",
    "        for new_record in record_list:\n",
    "            SeqIO.write(new_record, handle, \"fasta\")\n",
    "\n",
    "    # record the num of pre_binding for each gene\n",
    "    pre_binding_num[id] = pre_binding_num_tmp\n",
    "\n",
    "with open(tmp+pre_binding_num_file, \"w\") as f:\n",
    "    json.dump(pre_binding_num, f)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Blast and dataprocess\n",
    "NCBIXML: https://homolog.us/Biopython/Bio.Blast.NCBIXML.html#read/0\n",
    "\n",
    "BlastRecord: https://biopython.org/docs/1.75/api/Bio.Blast.Record.html\n",
    "\n",
    "XMLReader: https://codebeautify.org/xmlviewer# "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(file_out_dir+total_pre_binding_file_name, 'r') as f:\n",
    "    fasta_string = f.read()\n",
    "txid = [2697049]  # organism\n",
    "# Submit BLAST search and get handle object\n",
    "handle = NCBIWWW.qblast(program='blastn', megablast=\"yes\",\n",
    "                        database='refseq_rna', sequence=fasta_string,\n",
    "                        url_base='https://blast.ncbi.nlm.nih.gov/Blast.cgi', \n",
    "                        format_object='Alignment',\n",
    "                        format_type='Xml')\n",
    "# read handle object and save to a file\n",
    "with open(tmp+blast_results_file, 'w') as f:\n",
    "    f.write(handle.read())"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Decode the interested information from blast_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "align_num = []\n",
    "\n",
    "# read the id/plus-minus part/align_num\n",
    "with open(tmp+blast_results_file, 'r') as blast_output:\n",
    "    blast_records = NCBIXML.parse(blast_output)\n",
    "    loca = 0\n",
    "    for blast_record in blast_records:\n",
    "        align_accession = []\n",
    "        align_descrip_list = []\n",
    "        # get align num of each binding site\n",
    "        length = len(blast_record.alignments)\n",
    "        align_num.append(length)\n",
    "        for i in range(length):\n",
    "            descrip = blast_record.descriptions[i].title.split('|')\n",
    "            # get accession and descrip of each align seq\n",
    "            align_accession.append(descrip[3])\n",
    "            align_descrip_list.append(descrip[-1])\n",
    "        \n",
    "        FOI.loc[loca, 'align_accession'] = '|'.join(str(_) for _ in align_accession)\n",
    "        # add align_descrip to df\n",
    "        FOI.loc[loca, 'align_descrip'] = '|'.join(str(_) for _ in align_descrip_list)\n",
    "        # get plus/minus of each align seq\n",
    "        p_m = [blast_record.alignments[_].hsps[0].frame[1] for _ in range(length)]\n",
    "        # add plus/minus to df\n",
    "        try: FOI.loc[loca, 'plus/minus'] = ','.join([str(_) for _ in p_m])\n",
    "        except: FOI.loc[loca, 'plus/minus'] = 'NAN'\n",
    "        \n",
    "        loca += 1\n",
    "\n",
    "FOI['align_num'][:len(align_num)] = align_num"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the accession/gene_name/organism of alignments \n",
    "fetch_per_round = 50\n",
    "round = len(align_id_list) // fetch_per_round + 1\n",
    "for i in range(round):\n",
    "    id_list_per_round = align_id_list[i*fetch_per_round: (i+1)*fetch_per_round]\n",
    "    Entrez.email = \"1418767067@qq.com\"\n",
    "    handle = Entrez.efetch(db=\"nuccore\", strand=1,  # plus if strand=1\n",
    "                        id=id_list_per_round, rettype=\"gbwithparts\", retmode=\"text\")\n",
    "    seq_record = handle.read()\n",
    "    handle.close()\n",
    "    print('%1f%' % (i+1)/round*100)\n",
    "    with open(tmp+align_seq_in_file, 'a') as f:\n",
    "        f.write(seq_record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "align_accession, align_gene, align_mol_type, align_organism = [], [], [], []\n",
    "for align_seq_in in SeqIO.parse(tmp+align_seq_in_file, 'genbank'):\n",
    "    # get information of gene\n",
    "    align_accession.append(align_seq_in.id) # get seq id\n",
    "    # get gene_name\n",
    "    if align_seq_in.features:\n",
    "        for feature in align_seq_in.features:\n",
    "            if feature.type == \"CDS\":\n",
    "                gene_name = feature.qualifiers.get(\"gene\", ['NAN'])[0]\n",
    "    align_gene.append(gene_name)\n",
    "    # get molecule_type\n",
    "    align_mol_type.append(align_seq_in.annotations['molecule_type'])\n",
    "    \n",
    "    # get organism\n",
    "    align_organism.append(align_seq_in.annotations['organism'])\n",
    "\n",
    "st_pos = 0\n",
    "for i in range(len(FOI)):\n",
    "    align_num_tmp = align_num[i]\n",
    "    FOI.loc[i, ['align_accession']] \\\n",
    "             = [','.join(align_accession[st_pos: st_pos+align_num_tmp]), ','.join(align_gene[st_pos: st_pos+align_num_tmp]), \n",
    "                ','.join(align_mol_type[st_pos: st_pos+align_num_tmp]), ','.join(align_organism[st_pos: st_pos+align_num_tmp])]\n",
    "    st_pos += align_num_tmp"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Select wanted binding site "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "FOI['wanted'] = [True]*len(FOI)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sieve for the suitable binding site\n",
    "gene_name_list_out = [i for i in gene_name_list]\n",
    "for i in range(len(FOI)):\n",
    "    # check gene_name\n",
    "    gene_name = FOI.loc[i, 'gene_name']\n",
    "    if gene_name not in gene_name_list:\n",
    "        FOI.loc[i, 'wanted'] = False\n",
    "    else:\n",
    "        try: gene_name_list_out.remove(gene_name)\n",
    "        except: pass\n",
    "\n",
    "    # check DNA or mRNA type\n",
    "    if FOI.loc[i, 'wanted'] == True:\n",
    "        if FOI.loc[i, 'mol_type'] != 'mRNA':\n",
    "            FOI.loc[i, 'wanted'] = False\n",
    "        \n",
    "    # check gene_organism name\n",
    "    if FOI.loc[i, 'wanted'] == True:\n",
    "        spe_ori, gene_ori = FOI.loc[i, 'organism'], FOI.loc[i, 'gene_name']        \n",
    "        descrip = FOI.loc[i, 'align_descrip'].split('|')\n",
    "        for des in descrip:\n",
    "            if gene_ori not in des and spe_ori in des:\n",
    "                FOI.loc[i, 'wanted'] = False\n",
    "                break\n",
    "\n",
    "    # check plus/minus\n",
    "    if FOI.loc[i, 'wanted'] == True:\n",
    "        if pd.isnull(FOI.loc[i, 'plus/minus']):\n",
    "            FOI.loc[i, 'wanted'] = False\n",
    "        else:    \n",
    "            pm_list = FOI.loc[i, 'plus/minus'].split(',')\n",
    "            if '-1' not in pm_list:\n",
    "                FOI.loc[i, 'wanted'] = False\n",
    "    \n",
    "    # check G 40%-70%, non consective 5 base\n",
    "    if FOI.loc[i, 'wanted'] == True:\n",
    "        seq = FOI.loc[i, 'binding']\n",
    "        if 'GGGGG' in seq:\n",
    "            FOI.loc[i, 'wanted'] = False\n",
    "        else:\n",
    "            G_per = seq.count('G')/len(seq)\n",
    "            if G_per < 0.3 or G_per > 0.7:\n",
    "                FOI.loc[i, 'wanted'] = False\n",
    "\n",
    "# write the gene name to search next round\n",
    "with open(output+gene_name_list_tosearch, 'w') as f:\n",
    "    f.write('\\n'.join(gene_name_list_out))\n",
    "\n",
    "# write the whole information of interest to a excel file in tmp dir\n",
    "FOI.to_excel(tmp+'probes_sieve.xlsx')\n",
    "\n",
    "# get the sub dataframe of wanted probes\n",
    "output_df = FOI[FOI['wanted'] == True]\n",
    "# write the output to a xlsx file\n",
    "output_df.to_excel(output+'probes_wanted.xlsx')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
